core/src/main/java/io/questdb/cutlass/line/http/LineHttpSender.java

```diff
@@ -0,0 +1,637 @@
+                        client.disconnect();
+                    }
+                    break;
+                }
+                if (isRetryableHttpStatus(statusCode)) {
+                    long nowNanos = System.nanoTime();
+                    retryingDeadlineNanos = (retryingDeadlineNanos == Long.MIN_VALUE && !closing) ? nowNanos + maxRetriesNanos : retryingDeadlineNanos;
+                    if (nowNanos >= retryingDeadlineNanos) {
+                        handleHttpErrorResponse(statusCode, response);
+                    }
+                    client.disconnect(); // forces reconnect, just in case
+                    retryBackoff = backoff(retryBackoff);
+                    continue;
+                }
+                handleHttpErrorResponse(statusCode, response);
+            } catch (HttpClientException e) {
+                // this is a network error, we can retry
+                client.disconnect(); // forces reconnect
+                long nowNanos = System.nanoTime();
+                retryingDeadlineNanos = (retryingDeadlineNanos == Long.MIN_VALUE && !closing) ? nowNanos + maxRetriesNanos : retryingDeadlineNanos;
+                if (nowNanos >= retryingDeadlineNanos) {
+                    // we did our best, give up
+                    pendingRows = 0;
+                    request = newRequest();
+                    throw new LineSenderException("Could not flush buffer: ").put(url).put(" Connection Failed").put(": ").put(e.getMessage());
+                }
+                retryBackoff = backoff(retryBackoff);
+            }
+        }
+        pendingRows = 0;
+        request = newRequest();
+    }
+
+    private void handleHttpErrorResponse(DirectUtf8Sequence statusCode, HttpClient.ResponseHeaders response) {
+        // be ready for next request
+        pendingRows = 0;
+        request = newRequest();
+
+        CharSequence statusAscii = statusCode.asAsciiCharSequence();
+        if (Chars.equals("404", statusAscii)) {
+            consumeChunkedResponse(response);
+            client.disconnect();
+            throw new LineSenderException("Could not flush buffer: HTTP endpoint does not support ILP. [http-status=404]");
+        }
+        if (Chars.equals("401", statusAscii) || Chars.equals("403", statusAscii)) {
+            StringSink sink = Misc.getThreadLocalSink();
+            chunkedResponseToSink(response, sink);
+            LineSenderException ex = new LineSenderException("Could not flush buffer: HTTP endpoint authentication error");
+            if (sink.length() > 0) {
+                ex = ex.put(": ").put(sink);
+            }
+            ex.put(" [http-status=").put(statusAscii).put(']');
+            client.disconnect();
+            throw ex;
+        }
+        DirectUtf8Sequence contentType = response.getContentType();
+        if (contentType != null && Chars.equals("application/json", contentType.asAsciiCharSequence())) {
```

Comment:
nit: we have `Utf8s#equalsAscii` method - this way the code would be slightly more compact.